### PR TITLE
[SYCL-MLIR][NFC] Refactor code visiting arithmetic binary operators

### PR DESCRIPTION
**Pull request by victor-eds**
Each arithmetic binary operator is now handled outside the VisitBinaryOperator function in its own EmitBin(OP) function. This will allow reusing this code to implement compound assignment operators.

EmitBinOps ensures the LHS will be evaluated before the RHS.

Signed-off-by: Victor Perez <victor.perez@codeplay.com>

### CODE REVIEW

**Comment by etiotto** [comment]
In the description you mean: "EmitBinOps ensures the LHS will be evaluated **after** the RHS."

**Comment by victor-eds** [comment]
> In the description you mean: "EmitBinOps ensures the LHS will be evaluated **after** the RHS."

For the operations we're handling here, the LHS should be evaluated before the RHS, as it is in Clang

**Comment by etiotto** [comment]
> > In the description you mean: "EmitBinOps ensures the LHS will be evaluated **after** the RHS."
> 
> For the operations we're handling here, the LHS should be evaluated before the RHS, as it is in Clang

Yeah sorry, I see now in the test cases the type of binary ops you handle (e.g. x[++i] << y[i]).

File: polygeist/tools/cgeist/Test/Verification/binop.c [hunk 1]
```diff
@@ -0,0 +1,52 @@
+// RUN: cgeist %s --function=* -S -o - | FileCheck %s
+
+// COM: Checking evaluation order of binary operations.
+// COM: Each operation must be further tested in separate files.
+
+// CHECK-LABEL: func.func @f0(%arg0: memref<?xi32>, %arg1: memref<?xi32>) -> i32 attributes {llvm.linkage = #llvm.linkage<external>} {
+// CHECK-NEXT:    %0 = affine.load %arg0[0] : memref<?xi32>
+// CHECK-NEXT:    %1 = affine.load %arg1[1] : memref<?xi32>
```
[the remark this answers is not in the document]
Comment (victor-eds) [reply polygeist/tools/cgeist/Test/Verification/binop.c:8]: As we are testing evaluation order, I'd rather keep them as is. We want the order to be exactly that, right?

File: polygeist/tools/cgeist/Test/Verification/binop.c [hunk 2]
```diff
@@ -0,0 +1,52 @@
+// RUN: cgeist %s --function=* -S -o - | FileCheck %s
+
+// COM: Checking evaluation order of binary operations.
+// COM: Each operation must be further tested in separate files.
+
+// CHECK-LABEL: func.func @f0(%arg0: memref<?xi32>, %arg1: memref<?xi32>) -> i32 attributes {llvm.linkage = #llvm.linkage<external>} {
```
[the remark this answers is not in the document]
Comment (victor-eds) [reply polygeist/tools/cgeist/Test/Verification/binop.c]: Done

**Comment by victor-eds** [comment]
@etiotto I simply copied the code from each `case` in the original `VisitBinaryExpression` function to a new `EmitBinOp` function, that's why I did not format (also to make it easier for reviewers to see that's a copy). As we've discussed in a previous PR, some of these operators need to be revisited, as we're probably doing some things wrong. I am planning on doing so after putting up a PR fixing compound assignment operators evaluation order, so, if you're OK with that, we can postpone this refactoring till then.

**Review by whitneywhtsang — APPROVED** [review]
LGTM, am I correct that this is a NFCI?

**Comment by victor-eds** [comment]
> LGTM, am I correct that this is a NFCI?

It is. Control flow to calls to `Visit(Expr)` (being `Expr` is one of the expressions handled here) is changed also, as `Visit` first checks if there is a `VisitOPKIND` function present for each operation and calls `VisitBinaryOperator` otherwise.

File: polygeist/tools/cgeist/Test/Verification/binop.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (etiotto) [reply polygeist/tools/cgeist/Test/Verification/binop.c:8]: The 2 affine loads are independents, so they can appear in any order (with respect to each other).

**Comment by victor-eds** [comment]
@etiotto, I've addressed some of your comments. Not refactored the code, as this is yet to be revisited and will probably need to heavily modify it.

**Review by etiotto — APPROVED** [review]

**Pull request merged**